### PR TITLE
feat: Show image information in UI (#10486)

### LINKED ISSUE #10486: Show more information about images in the UI
# Summary

When I click on a pod, one of the key pieces of information I see is the image. This is sometimes the only indication of what code the app is running. I want to see more information.

# Motivation

To make it easier to understand what an app is doing.

# Proposal

Image meta-data can be queried from the repository, like Argo Workflows does:

https://github.com/argoproj/argo-workflows/blob/26ab0aed8ba19571ffe3a2b048fcb43cbd1986e3/workflow/controller/entrypoint/container_registery_index.go#L17

This is the same as running `docker image inspect`, but this is pretty basic: entry point, env vars. 

I'd like to learn how the image built, so maybe linking to hub might be good:

https://hub.docker.com/r/argoproj/argocd

Also, `docker image history` shows interesting information:

 ```
IMAGE          CREATED       CREATED BY                                      SIZE      COMMENT
6e7cfb90c9e7   6 hours ago   USER 999                                        0B        buildkit.dockerfile.v0
<missing>      6 hours ago   RUN /bin/sh -c ln -s /usr/local/bin/argocd /…   168B      buildkit.dockerfile.v0
<missing>      6 hours ago   USER root                                       0B        buildkit.dockerfile.v0
<missing>      6 hours ago   COPY /go/src/github.com/argoproj/argo-cd/dis…   134MB     buildkit.dockerfile.v0
<missing>      7 hours ago   WORKDIR /home/argocd                            0B        buildkit.dockerfile.v0
<missing>      7 hours ago   USER 999                                        0B        buildkit.dockerfile.v0
<missing>      7 hours ago   ENV USER=argocd                                 0B        buildkit.dockerfile.v0
<missing>      7 hours ago   RUN /bin/sh -c mkdir -p tls &&     mkdir -p …   0B        buildkit.dockerfile.v0
<missing>      7 hours ago   WORKDIR /app/config                             0B        buildkit.dockerfile.v0
<missing>      7 hours ago   RUN /bin/sh -c touch ssh_known_hosts &&     …   31B       buildkit.dockerfile.v0
<missing>      7 hours ago   WORKDIR /app/config/ssh                         0B        buildkit.dockerfile.v0
<missing>      7 hours ago   RUN /bin/sh -c ln -s /usr/local/bin/entrypoi…   28B       buildkit.dockerfile.v0
<missing>      7 hours ago   COPY entrypoint.sh /usr/local/bin/entrypoint…   205B      buildkit.dockerfile.v0
<missing>      7 hours ago   COPY /usr/local/bin/kustomize /usr/local/bin…   14.4MB    buildkit.dockerfile.v0
<missing>      7 hours ago   COPY /usr/local/bin/helm /usr/local/bin/helm…   46.4MB    buildkit.dockerfile.v0
<missing>      7 hours ago   COPY hack/git-verify-wrapper.sh /usr/local/b…   934B      buildkit.dockerfile.v0
<missing>      7 hours ago   COPY hack/gpg-wrapper.sh /usr/local/bin/gpg-…   215B      buildkit.dockerfile.v0
<missing>      7 hours ago   RUN /bin/sh -c groupadd -g $ARGOCD_USER_ID a…   102MB     buildkit.dockerfile.v0
<missing>      7 hours ago   ENV DEBIAN_FRONTEND=noninteractive              0B        buildkit.dockerfile.v0
<missing>      7 hours ago   ENV ARGOCD_USER_ID=999                          0B        buildkit.dockerfile.v0
<missing>      7 hours ago   USER root                                       0B        buildkit.dockerfile.v0
<missing>      4 weeks ago   /bin/sh -c #(nop)  CMD ["bash"]                 0B
<missing>      4 weeks ago   /bin/sh -c #(nop) ADD file:396eeb65c8d737180…   77.8M
```

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #10486

image

### CODE REVIEW

File: server/image/image.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type server struct {
+	keychain authn.Keychain
+}
+
+func (s *server) GetImage(ctx context.Context, req *imagepkg.GetImageRequest) (*imagepkg.GetImageResponse, error) {
+	log.WithField("image", req.Image).Info("GetImage")
```
Comment (crenshaw-dev) [inline server/image/image.go:21]: Would want to truncate the image name before logging.

File: server/image/image.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type server struct {
+	keychain authn.Keychain
+}
+
+func (s *server) GetImage(ctx context.Context, req *imagepkg.GetImageRequest) (*imagepkg.GetImageResponse, error) {
```
Comment (crenshaw-dev) [inline server/image/image.go:20]: What would you think are the security considerations associated with this request? Basically any authenticated user gets to make this request?

File: server/image/image.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	log.WithField("image", req.Image).Info("GetImage")
+	ref, err := name.ParseReference(req.Image)
+	if err != nil {
+		return nil, err
+	}
+	img, err := remote.Image(ref, remote.WithAuthFromKeychain(s.keychain))
```
Comment (crenshaw-dev) [inline server/image/image.go:26]: Would we want to be able to support more than one container registry, e.g. a public one and then one or more private ones (e.g. a test and a prod internal registry)?

File: ui/src/app/applications/components/application-node-info/container-image-details.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    const [resp, setResp] = useState<GetImageResponse>();
+    useEffect(() => {
+        services.images
+            .get(image)
+            .then(setResp)
+            .catch(setError);
+    }, [image]);
+
+    const parts = image.replace(/:.*/, '').split('/');
+    const repo = (parts.length === 1 ? '_/' : '') + parts.join('/');
```
Comment (crenshaw-dev) [inline ui/src/app/applications/components/application-node-info/container-image-details.tsx:19]: Not sure I follow this bit... I feel like we oughta use the `URL` object to manipulate URL-like things.

File: server/image/image.go (hunk 3 above)
Comment (crenshaw-dev) [inline server/image/image.go:26]: I think we'd want a registry allowlist to avoid making requests to arbitrary hosts.

**Review by crenshaw-dev — COMMENTED** [review]
This is cool.

Personally I still feel like I'd like to see this as a use case for proxy-type UI extensions. 

cc: @leoluz @zachaller for y'alls thoughts.

Considering this use case may help us think through the RBAC design for proxy UI extensions.

File: server/image/image.go (hunk 2 above)
Comment (alexec) [reply server/image/image.go:20]: I had considered this. I'd worked on the assumption you could inspect any image from the repo, but maybe we should only allow you to inspect images for apps you have access to.

File: ui/src/app/applications/components/application-node-info/container-image-details.tsx (hunk 4 above)
Comment (alexec) [reply ui/src/app/applications/components/application-node-info/container-image-details.tsx:19]: This is a hack create the right URL for hub.docker.com when the image is does no contain slashes, you need to put "_/" into the URL..

File: server/image/image.go (hunk 3 above)
Comment (alexec) [reply server/image/image.go:26]: You can't do this, you must configure Argo CD service with your image repo. Only admin can do that.

Comment (ashutosh16) [reply server/image/image.go:26]: How do we connect to the private repository? How is the keychain populated?

File: server/image/image.go (hunk 2 above)
Comment (alexec) [reply server/image/image.go:20]: At my company, I think you can pull any image from our private repository. I'm not saying that's a good idea, just that engineers already have that ability.

File: server/image/image.go (hunk 1 above)
Comment (alexec) [reply server/image/image.go:21]: I will delete this line.

**Pull request closed without merging**